### ISSUE TITLE
Unable to use transactions with Azure Synapse Analytics

### ISSUE THREAD

**Issue by gordthompson** [sqlalchemy/sqlalchemy#8231]
### Discussed in https://github.com/sqlalchemy/sqlalchemy/discussions/8223

<div type='discussions-op-text'>

<sup>Originally posted by **ww2406** July  7, 2022</sup>
### Describe the bug

Hello!

**Problem**: Unable to use transactions with Azure Synapse Analytics

**Description**: ODBC Error 111214 is returned: An attempt to complete a transaction has failed. No corresponding transaction found.

**Other notes**: Transactions function properly when using pyodbc directly [see MCVE]. I saw [this](https://stackoverflow.com/a/69472955) post by gordthompson in response to a question on StackOverflow (and since added to the documentation) that

> Azure SQL Data Warehouse does not support transactions

but this is outdated. Azure Synapse Analytics does support transactions in most cases; see [this](https://docs.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/sql-data-warehouse-develop-transactions) Microsoft documentation:

> As you would expect, SQL pool supports transactions as part of the data warehouse workload. However, to ensure SQL pool is maintained at scale, some features are limited when compared to SQL Server. This article highlights the differences.

### To Reproduce

```python
import pyodbc 
import sqlalchemy as sa 
import os 

p_con_string = os.getenv("synapse_conn_pyodbc")
sa_eng_string = os.getenv("synapse_conn_sa")

p_con = pyodbc.connect(p_con_string)
engine = sa.create_engine(sa_eng_string)

p_cur = p_con.cursor()
p_cur.execute("INSERT INTO test VALUES ('test')")
p_con.commit()
p_con.close()
# No error

with engine.begin() as s_con:
    stmt = sa.Text("INSERT INTO test VALUES ('test')")
    s_con.execute(stmt)

# Traceback (most recent call last):
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3280, in _wrap_pool_connect
#     return fn()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 310, in connect
#     return _ConnectionFairy._checkout(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 868, in _checkout
#     fairy = _ConnectionRecord.checkout(pool)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 476, in checkout
#     rec = pool._do_get()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 145, in _do_get
#     with util.safe_reraise():
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
#     compat.raise_(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
#     raise exception
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 143, in _do_get
#     return self._create_connection()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 256, in _create_connection
#     return _ConnectionRecord(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 371, in __init__
#     self.__connect()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 679, in __connect
#     )._exec_w_sync_on_first_run(self.dbapi_connection, self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 329, in _exec_w_sync_on_first_run
#     self(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 343, in __call__
#     fn(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 1695, in go
#     return once_fn(*arg, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\create.py", line 694, in first_connect
#     dialect.do_rollback(c.connection)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\default.py", line 682, in do_rollback
#     dbapi_connection.rollback()
# pyodbc.ProgrammingError: ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]111214;An attempt to complete a transaction has failed. No corresponding transaction found. (111214) (SQLEndTran)')

# The above exception was the direct cause of the following exception:

# Traceback (most recent call last):
#   File "synapse_issue.py", line 15, in <module>
#     with engine.begin() as s_con:
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3062, in begin
#     conn = self.connect(close_with_result=close_with_result)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3234, in connect
#     return self._connection_cls(self, close_with_result=close_with_result)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 96, in __init__
#     else engine.raw_connection()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3313, in raw_connection
#     return self._wrap_pool_connect(self.pool.connect, _connection)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3283, in _wrap_pool_connect
#     Connection._handle_dbapi_exception_noconnection(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 2117, in _handle_dbapi_exception_noconnection
#     util.raise_(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
#     raise exception
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3280, in _wrap_pool_connect
#     return fn()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 310, in connect
#     return _ConnectionFairy._checkout(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 868, in _checkout
#     fairy = _ConnectionRecord.checkout(pool)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 476, in checkout
#     rec = pool._do_get()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 145, in _do_get
#     with util.safe_reraise():
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
#     compat.raise_(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
#     raise exception
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 143, in _do_get
#     return self._create_connection()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 256, in _create_connection
#     return _ConnectionRecord(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 371, in __init__
#     self.__connect()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 679, in __connect
#     )._exec_w_sync_on_first_run(self.dbapi_connection, self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 329, in _exec_w_sync_on_first_run
#     self(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 343, in __call__
#     fn(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 1695, in go
#     return once_fn(*arg, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\create.py", line 694, in first_connect
#     dialect.do_rollback(c.connection)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\default.py", line 682, in do_rollback
#     dbapi_connection.rollback()
# sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]111214;An attempt to complete a transaction has failed. No 
# corresponding transaction found. (111214) (SQLEndTran)')
# (Background on this error at: https://sqlalche.me/e/14/f405)
```

### Error

```
# Traceback (most recent call last):
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3280, in _wrap_pool_connect
#     return fn()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 310, in connect
#     return _ConnectionFairy._checkout(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 868, in _checkout
#     fairy = _ConnectionRecord.checkout(pool)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 476, in checkout
#     rec = pool._do_get()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 145, in _do_get
#     with util.safe_reraise():
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
#     compat.raise_(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
#     raise exception
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 143, in _do_get
#     return self._create_connection()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 256, in _create_connection
#     return _ConnectionRecord(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 371, in __init__
#     self.__connect()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 679, in __connect
#     )._exec_w_sync_on_first_run(self.dbapi_connection, self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 329, in _exec_w_sync_on_first_run
#     self(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 343, in __call__
#     fn(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 1695, in go
#     return once_fn(*arg, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\create.py", line 694, in first_connect
#     dialect.do_rollback(c.connection)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\default.py", line 682, in do_rollback
#     dbapi_connection.rollback()
# pyodbc.ProgrammingError: ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]111214;An attempt to complete a transaction has failed. No corresponding transaction found. (111214) (SQLEndTran)')

# The above exception was the direct cause of the following exception:

# Traceback (most recent call last):
#   File "synapse_issue.py", line 15, in <module>
#     with engine.begin() as s_con:
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3062, in begin
#     conn = self.connect(close_with_result=close_with_result)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3234, in connect
#     return self._connection_cls(self, close_with_result=close_with_result)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 96, in __init__
#     else engine.raw_connection()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3313, in raw_connection
#     return self._wrap_pool_connect(self.pool.connect, _connection)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3283, in _wrap_pool_connect
#     Connection._handle_dbapi_exception_noconnection(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 2117, in _handle_dbapi_exception_noconnection
#     util.raise_(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
#     raise exception
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\base.py", line 3280, in _wrap_pool_connect
#     return fn()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 310, in connect
#     return _ConnectionFairy._checkout(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 868, in _checkout
#     fairy = _ConnectionRecord.checkout(pool)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 476, in checkout
#     rec = pool._do_get()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 145, in _do_get
#     with util.safe_reraise():
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
#     compat.raise_(
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\compat.py", line 208, in raise_
#     raise exception
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\impl.py", line 143, in _do_get
#     return self._create_connection()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 256, in _create_connection
#     return _ConnectionRecord(self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 371, in __init__
#     self.__connect()
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\pool\base.py", line 679, in __connect
#     )._exec_w_sync_on_first_run(self.dbapi_connection, self)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 329, in _exec_w_sync_on_first_run
#     self(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\event\attr.py", line 343, in __call__
#     fn(*args, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\util\langhelpers.py", line 1695, in go
#     return once_fn(*arg, **kw)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\create.py", line 694, in first_connect
#     dialect.do_rollback(c.connection)
#   File "C:\Users\user\AppData\Roaming\Python\Python310\site-packages\sqlalchemy\engine\default.py", line 682, in do_rollback
#     dbapi_connection.rollback()
# sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]111214;An attempt to complete a transaction has failed. No 
# corresponding transaction found. (111214) (SQLEndTran)')
# (Background on this error at: https://sqlalche.me/e/14/f405)
``` 

### Versions

- OS: Windows 10
- Python: 3.10
- SQLAlchemy: 1.4.37
- Database: Azure Synapse Analytics Dedicated Pool
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): pyodbc

### Additional context

_No response_</div>

**Comment by zzzeek** [comment]
@gordthompson can you please elaborate what the actual SQLAlchemy bug is?   the trace shows pyodbc.connection.rollback() failing.  how is that our issue ?

**Comment by gordthompson** [comment]
When we `get_isolation_level()` we have to look in two places:

https://github.com/sqlalchemy/sqlalchemy/blob/805a1323b973a30af99ce506dd5c5c4ab96cff0f/lib/sqlalchemy/dialects/mssql/base.py#L2927

Plain old SQL Server uses the former, while the Data Warehouse products use the latter.

@ww2406 [discovered](https://github.com/sqlalchemy/sqlalchemy/discussions/8223#discussioncomment-3109049) that if they removed the first entry and just queried `sys.dm_pdw_nodes_exec_sessions` then the error from Azure Synapse Analytics went away. Subsequent testing showed that simply swapping the order of the views made Synapse happy and SQL Server 2019 was cool with it. (I ran `tox -e py38-mssql` just to be sure.)

So this is just a workaround for the difference in behaviour between SQL Server and Synapse.

**Comment by zzzeek** [comment]
OK I need more detail than that:

1. has this code *ever* worked for data warehouse?
2. if so, what has changed?
3. is it the case that both sys.dm_exec_sessions and sys.dm_pdw_nodes_exec_sessions both exist in data warehouse products?
4. if so, what's the difference between them?
5. does the sys.dm_pdw_nodes_exec_sessions view exist in plain SQL server ? because if it does, and returns garbage data as seems to be suggested by number 3 above, then it would be a disaster for us to reverse these
6. can we change the entire approach and look at SQL server version string, or some other reliable indicator, rather than just querying both tables and waiting for a failure?

tl;dr; reversing them is a band aid, lets please use real server telemetry to choose the correct view up front

**Comment by zzzeek** [comment]
if the "reversing" helps only because it avoids an error condition, that otherwise somehow corrupts the connection, then that also is kind of a bigger problem.

**Comment by CaselIT** [comment]
Are the tables in sys reported by the information tables, like `sys.tables`? Because if they are maybe we can select the table to use based on that?

**Comment by gordthompson** [comment]
> Are the tables in sys reported by the information tables, like `sys.tables`? Because if they are maybe we can select the table to use based on that?

Yes, at least on SQL Server 2019:

```
1> select count(*) as n from sys.system_views where name = 'dm_exec_sessions'
2> go
n          
-----------
          1

(1 rows affected)
1> select count(*) as n from sys.system_views where name = 'dm_pdw_nodes_exec_sessions'
2> go
n          
-----------
          0

(1 rows affected)
```

**Comment by zzzeek** [comment]
yes I like looking in system tables for it.  

however, I really need to understand the actual problem.   

is the problem the following?

1. we try to query nonexistent dm_exec_sessions
2. error is raised, we catch
3. ....  connection is corrupted?  can't rollback ?   isn't this a bigger problem for issues like determining connection invalidation?  should we really just be fixing [is_disconnect](https://github.com/sqlalchemy/sqlalchemy/blob/805a1323b973a30af99ce506dd5c5c4ab96cff0f/lib/sqlalchemy/dialects/mssql/pyodbc.py#L675)?

what's involved getting a connection to one of these DBs, is there some msft developer cloud thing we can try?

**Comment by zzzeek** [comment]
I need the answser to number 3 above, but also if someone wants to work on a "test just the correct table only" patch, that is fine as well

**Comment by zzzeek** [comment]
OK @ww2406 has added a [comment](https://github.com/sqlalchemy/sqlalchemy/pull/8233#issuecomment-1179643328) showing the original, actual bug, which is what it looked like from the beginning, which is pyodbc is being put into an invalid state it does not expect:

```py
import pyodbc

conn_string = ''
con = pyodbc.connect(conn_string)
cur = con.cursor()
cur.execute("SELECT * FROM sys.dm_exec_sessions")
# ProgrammingError -- catalog view is not supported in this version
con.rollback()
# 111214 -- attempt to complete transaction failed
```

Here's why this is important:

1. it's a bug in pyodbc or related drivers and should be reported, or noted, or somehow understood at the driver level.
2. it implies that **maybe all, or many more** rollback calls will not work on this database / driver combination ?

**Comment by gordthompson** [comment]
@ww2406 - Perhaps Synapse might be deferring the start of the transaction?

If you create the stored procedure

```sql
CREATE PROCEDURE [dbo].[sqla_issue_8231] 
AS
DECLARE @n int;
BEGIN
    SET NOCOUNT ON;

    SELECT @n = COUNT(*) FROM sys.system_views WHERE name = 'dm_exec_sessions';
    -- purposely try hitting the view that does not exist
    IF @n = 1
        SELECT COUNT(*) AS i FROM sys.dm_pdw_nodes_exec_sessions;
    ELSE
        SELECT COUNT(*) AS i FROM sys.dm_exec_sessions;
END
```

and then run the following script in SSMS

```sql
BEGIN TRANSACTION
BEGIN TRY
    EXEC dbo.sqla_issue_8231;
END TRY
BEGIN CATCH
    -- ignore error
END CATCH
ROLLBACK
```

does Synapse throw an error?

**Comment by zzzeek** [comment]
well yes the failure is related to the fact that a SQL statement fails, which is exactly when youd *want* to call rollback, and then the rollback fails, because somehow that SQL statement's error condition has *already* rolled back the transaction, pyodbc doesnt know about it, and boom.  this is all wrong on someone's end.

**Comment by zzzeek** [comment]
my point is the same as always, DBAPI is not supposed to have restrictions on rollback(), otherwise you'd never know when you can call it or not if ceratin error conditions "autorollback" and break the driver's knowledge of the sequence of events.  

this is where pyodbc pretending to be "database agnostic" (95% of all pyodbc use is for SQL server) is a problem because it really should have logic to know that certain error conditions mean the driver just did an "autorollback", unless there's some way to configure this with the odbc driver or similar.

**Comment by zzzeek** [comment]
I would not feel comfortable squashing an exception like that for something as critical as a database rollback.

better would be if there were logic that literally ran "select can_we_rollback_here", whatever that has to be, then does the rollback.

*then*, we make it a recipe for those using Azure Synapse.  I dont want hardcoded logic in the dialect that adds latency to all MSSQL code everywhere to work around what is IMO a major bug in Synapse and/or drivers.

**Comment by zzzeek** [comment]
Except we don't at the moment have a DialectEvents.do_rollback() method that can conditionilalize whether rollback is called or not.

So to be clear we have fully confirmed there is aboslutely no way to get the DB or ODBC driver to not do this "autorollback", and what would pyodbc devs actually tell us to do ?  catch all exceptions and regexp them ?

**Comment by zzzeek** [comment]
> If an error other than a run-time statement error prevents the successful completion of an explicit transaction, Azure Synapse Analytics automatically rolls back the transaction and frees all resources held by the transaction. F

the error that we are catching **is a run-time statement error**.  this is a bug in synapse

**Comment by zzzeek** [comment]
> If Synapse starts a transaction at the client's request and later rolls it back arbitrarily for whatever reason then it shouldn't raise an error if the client requests the rollback that has already occurred.

the [doc](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/transactions-sql-data-warehouse?view=aps-pdw-2016-au7#error-handling) is explicit that it will:

> If COMMIT or ROLLBACK are run and there is no active transaction, an error is raised.

**Comment by gordthompson** [comment]
> @zzzeek did you see this?
> 
> > If a run-time statement error occurs in a batch, Azure Synapse Analytics behaves consistent with SQL Server XACT_ABORT set to ON and the entire transaction is rolled back.

but even that is inconsistent between Synapse and SQL Server - ref: [here](https://github.com/sqlalchemy/sqlalchemy/issues/8231#issuecomment-1180462220)

**Comment by zzzeek** [comment]
can you reference "Fixes: #8231" in your PR so it links here, thanks

**Comment by gordthompson** [comment]
This is what I've been working on for xact_abort:

https://github.com/gordthompson/sqlalchemy/commit/de959b336e117dd5451fa728ed0a85872032fd53

I've tested it as much as I can without Synapse, but it looks like it should work.

**Comment by gordthompson** [comment]
> > This is what I've been working on for xact_abort:
> > [gordthompson@de959b3](https://github.com/gordthompson/sqlalchemy/commit/de959b336e117dd5451fa728ed0a85872032fd53)
> > I've tested it as much as I can without Synapse, but it looks like it should work.
> 
> great minds think alike, nice job

Thanks.

@ww2406 - Can you test that patch in isolation (i.e., without your select-o-matic patch to use the appropriate view) and verify that it works? That is

```
pip install git+https://github.com/gordthompson/sqlalchemy@xact_abort
```

(or equivalent) and then do something like

```python
import sqlalchemy as sa

engine = sa.create_engine(connection_url, xact_abort=True)

with engine.begin() as conn:
    pass
```

**Comment by gordthompson** [comment]
> @gordthompson sadly 111214 is still raised :(

Rats. Okay, thanks for checking.

**Comment by gordthompson** [comment]
Thanks, @ww2406 . Having a real DW connection certainly would be helpful to try and figure out what the heck is going on with this 111214 thing.

**Comment by gordthompson** [comment]
Okay, with big thanks to @ww2406 

https://github.com/gordthompson/sqlalchemy/tree/xact_abort_0721a

now has a patch that works for me against a real Synapse instance.

After doing

```
pip install git+https://github.com/gordthompson/sqlalchemy@xact_abort_0721a
```

the following code

```python
import logging

import sqlalchemy as sa

logging.basicConfig(level=logging.INFO)

connection_url = "<insert your connection URL here>"
engine = sa.create_engine(
    connection_url,
    xact_abort=True,
)

with engine.begin() as conn:
    pass
```

should not fail, but simply log

> INFO:root:pyodbc.ProgrammingError 111214 'No corresponding transaction found.' has been suppressed via xact_abort=True

If this looks okay then I can either
- try submitting an addendum to #8233 , or
- submit this as a separate change.

**Comment by zzzeek** [comment]
let's keep this all on one PR / gerrit if possible.

**Comment by zzzeek** [comment]
i thought gord could append to your patch.   either way.   sorry I just wait for the "it's done, here's our solution" patch to come out before I look too closely :)

**Comment by sqla-tester** [comment]
**Gord Thompson** has proposed a fix for this issue in the **main** branch:

**Fix 'No … transaction found' error on Synapse.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4006

**Comment by sqla-tester** [comment]
**Gord Thompson** has proposed a fix for this issue in the **rel_1_4** branch:

**Fix 'No transaction found' error on Synapse.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4020

**Comment by cwegener** [comment]
> Of course, if only a certain Redmond-based company that knows these products inside and out (hopefully) would consider writing a Python client like, idk, virtually every other database vendor…. man_shrugging

A bit OT: But there is at least a whole documentation section on docs.microsoft.com about pymssql/FreeTDS ... https://docs.microsoft.com/en-us/sql/connect/python/pymssql/step-1-configure-development-environment-for-pymssql-python-development?view=sql-server-ver16

**Comment by iamshaikarshad** [comment]
Hi Guys,

I initially i was able to connect to the syanpse dedicated pool and perform GET request, However was not able to perform POST request and received the same error.

I followed the above approach of:
`pip install git+https://github.com/gordthompson/sqlalchemy@xact_abort_0721a`
and
`params = urllib.parse.quote_plus("Driver={ODBC Driver 17 for SQL Server};Server=tcp:database.sql.azuresynapse.net,1433;Database=dedicatedpool;Uid=<Username>;Pwd=<Password>;Encrypt=yes;Connection Timeout=30;Authentication=ActiveDirectoryPassword;xact_abort=True")`

But after performing this action I was not evern able to connect with the server.

With error same as earlier.
`o_rollback
    dbapi_connection.rollback()
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]111214;An attempt to complete a transaction has failed. No corresponding transaction found. (111214) (SQLEndTran)')       
(Background on this error at: https://sqlalche.me/e/20/f405)`

Can you please advise me on this, if there are any additional steps to be performed?

**Comment by zzzeek** [comment]
looks like instructions for this are at https://docs.sqlalchemy.org/en/20/dialects/mssql.html#avoiding-transaction-related-exceptions-on-azure-synapse-analytics
